src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt

```diff
@@ -0,0 +1,216 @@
+            bodyAudio = body?.let { constructAssociatedAudio(body) }
+        )
+    }
+
+    private fun constructResources(content: Content): Observable<Resource> {
+        return resourceRepository
+            .getByContent(content)
+            .flattenAsObservable(this::contentsToResources)
+            .cache()
+    }
+
+    private fun constructResources(collection: Collection): Observable<Resource> {
+        return resourceRepository
+            .getByCollection(collection)
+            .flattenAsObservable(this::contentsToResources)
+            .cache()
+    }
+
+    private fun contentsToResources(contents: Iterable<Content>): Iterable<Resource> {
+        return contents
+            .sortedWith(compareBy({ it.start }, { it.sort }))
+            .zipWithNext { a, b ->
+                when {
+                    a.labelKey != "title" -> null
+                    b.labelKey != "body" -> constructResource(a, null)
+                    else -> constructResource(a, b)
+                }
+            }
+            .filterNotNull()
+    }
+
+    private fun workbookTake(modelTake: org.wycliffeassociates.otter.common.data.model.Take): Take {
+        // TODO: move into TakeMapper?
+        return Take(
+            name = modelTake.filename,
+            file = modelTake.path,
+            number = modelTake.number,
+            format = MimeType.WAV, // TODO
+            createdTimestamp = modelTake.timestamp,
+            deletedTimestamp = BehaviorRelay.createDefault(null) // TODO: need DB deleted field
+        )
+    }
+
+    private fun modelTake(workbookTake: Take): org.wycliffeassociates.otter.common.data.model.Take {
+        // TODO: move into TakeMapper?
+        return org.wycliffeassociates.otter.common.data.model.Take(
+            filename = workbookTake.file.name,
+            path = workbookTake.file,
+            number = workbookTake.number,
+            timestamp = workbookTake.createdTimestamp,
+            played = false,
+            markers = listOf()
+        )
+    }
+
+    private fun constructAssociatedAudio(content: Content): AssociatedAudio {
+        /** Map to recover model.Take objects from workbook.Take objects. */
+        val takeMap = WeakHashMap<Take, org.wycliffeassociates.otter.common.data.model.Take>()
+
+        /** Initial Takes read from the DB. */
+        val takesFromDb = takeRepository
+            .getByContent(content)
+            .flattenAsObservable { list -> list.sortedBy { it.number } }
+            .map { workbookTake(it) to it }
+
+        /** Relay to send Takes out to consumers, but also receive new Takes from UI. */
+        val takesRelay = ReplayRelay.create<Take>()
+        takesFromDb
+            // Record the mapping between data types.
+            .doOnNext { (wbTake, modelTake) -> takeMap[wbTake] = modelTake }
+            // Feed the initial list to takesRelay
+            .map { (wbTake, _) -> wbTake }
+            .subscribe(takesRelay)
```

Comment:
Trying to make sure I understand what's going on- am I right to say that the TakesRelay will receive a take from the DB (converted to a wbTake), and emit that take so that the UI can respond and display it? The TakeRelay will not do anything else with takes received from the DB since they will already be in the takeMap.

Reply:
@KJoslyn Yes, but not just "a take". It will be zero, one, or many takes. But as you said, they're only loaded from the DB once. Each relay might get updates from the UI layer, but never again from the DB.